com.oracle.truffle.r.ffi.impl/src/com/oracle/truffle/r/ffi/impl/altrep/AltrepDownCallNodeImpl.java

```diff
@@ -0,0 +1,152 @@
+        return AltrepDownCallNodeImplNodeGen.create();
+    }
+
+    public static AltrepDownCallNodeImpl getUncached() {
+        return AltrepDownCallNodeImplNodeGen.getUncached();
+    }
+
+    @Specialization(guards = "cachedLength == args.length", limit = "3")
+    public Object doIt(AltrepMethodDescriptor altrepDowncallIn, boolean unwrapResult, boolean[] wrapArguments, Object[] args,
+                    @Cached("args.length") int cachedLength,
+                    @CachedLibrary("altrepDowncallIn.method") InteropLibrary methodInterop,
+                    @Cached(value = "createUnwrapNode(unwrapResult)", uncached = "createUncachedUnwrapNode()") FFIUnwrapNode unwrapNode,
+                    @CachedContext(TruffleRLanguage.class) ContextReference<RContext> ctxRef,
+                    @Cached("createMaterialized(wrapArguments)") FFIMaterializeNode[] materializeNodes,
+                    @Cached("createToNatives(wrapArguments)") FFIToNativeMirrorNode[] toNativeNodes,
+                    @Cached("createBinaryProfile()") ConditionProfile isLLVMProfile,
+                    @Cached BranchProfile unwrapResultProfile,
+                    @Cached("createIdentityProfile()") ValueProfile identityProfile,
+                    @Cached AfterDownCallProfiles afterDownCallProfiles) {
+        CompilerAsserts.partialEvaluationConstant(unwrapResult);
+        AltrepMethodDescriptor altrepMethodDescriptor = identityProfile.profile(altrepDowncallIn);
+
+        assert methodInterop.isExecutable(altrepMethodDescriptor.method);
+        RContext ctx = ctxRef.get();
+
+        Object before = null;
+        if (isLLVMProfile.profile(altrepMethodDescriptor.rffiType == Type.LLVM)) {
+            before = ctx.getRFFI(TruffleLLVM_Context.class).beforeDowncall(null, altrepMethodDescriptor.rffiType);
+        } else {
+            before = ctx.getRFFI(TruffleNFI_Context.class).beforeDowncall(null, altrepMethodDescriptor.rffiType);
+        }
+
+        Object ret;
+        try (FFIDownCallWrap ffiWrap = new FFIDownCallWrap(cachedLength)) {
+            Object[] wrappedArgs = ffiWrap.wrapSome(args, materializeNodes, toNativeNodes, wrapArguments);
+            ret = methodInterop.execute(altrepMethodDescriptor.method, wrappedArgs);
+            if (unwrapResult) {
+                unwrapResultProfile.enter();
+                ret = unwrapNode.execute(ret);
+            }
+        } catch (Exception ex) {
+            throw RInternalError.shouldNotReachHere(ex);
+        }
+
+        if (isLLVMProfile.profile(altrepMethodDescriptor.rffiType == Type.LLVM)) {
+            ctx.getRFFI(TruffleLLVM_Context.class).afterDowncall(before, altrepMethodDescriptor.rffiType, afterDownCallProfiles);
+        } else {
+            ctx.getRFFI(TruffleNFI_Context.class).afterDowncall(before, altrepMethodDescriptor.rffiType, afterDownCallProfiles);
+        }
+
+        return ret;
+    }
+
+    @Specialization(guards = "cachedLength == args.length", replaces = "doIt")
```

Comment:
This is not going to work, but we probably never reach this fallback because the arguments count is always constant (right?). The first specialization can just `CompilerAsserts.partialEvaluationContant(args.length)` and it should not have the guard at all.

This issue with this fallback is that `FFIDownCallWrap` is not really ready to handle dynamically sized array of arguments (it has `@ExplodeLoop`), so the compiler will complain that the array is not constantly sized. Moreover, there should not be any guard on this `@Specialization`, because then it cannot work as a more generic replacement for the previous `@Specialization`

Reply:
I see. The argument count is always constant - for exactly the same reason why, e.g., `AltRepClassDescriptor.lengthMethodSignature` is constant. So can we have only the first specialization without any guards and with only `partialEvaluationConstant(args.length)` statement? I guess that we do not need any uncached fallback after all.

Reply:
yes, exactly